orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/WebhookResponseProcessor.java

```diff
@@ -0,0 +1,232 @@
+    Map<String, Object> webHookOutput = (Map<String, Object>) stageOutput.get("webhook");
+    webHookOutput.put("statusCode", response.getStatusCode());
+    webHookOutput.put("statusCodeValue", response.getStatusCode().value());
+
+    if (response.getBody() != null) {
+      webHookOutput.put("body", response.getBody());
+    }
+    HttpStatus status = response.getStatusCode();
+
+    if (status.is2xxSuccessful() || status.is3xxRedirection()) {
+
+      // Retrieve status check url
+      if (stageData.waitForCompletion != null && stageData.waitForCompletion) {
+        String statusUrl;
+        try {
+          statusUrl = new WebhookStatusCheckUrlRetriever().getStatusCheckUrl(response, stageData);
+        } catch (Exception e) {
+          webHookOutput.put(
+              "error", "Exception while resolving status check URL: " + e.getMessage());
+          log.error("Exception received while determining status check url", e);
+          return TaskResult.builder(ExecutionStatus.TERMINAL).context(stageOutput).build();
+        }
+
+        if (StringUtils.isEmpty(statusUrl)) {
+          webHookOutput.put(
+              "error",
+              "The status URL couldn't be resolved, but 'Wait for completion' was checked");
+          return TaskResult.builder(ExecutionStatus.TERMINAL).context(stageOutput).build();
+        }
+        stageData.statusEndpoint = statusUrl;
+        stageExecution.getContext().put("statusEndpoint", statusUrl);
+        webHookOutput.put("statusEndpoint", statusUrl);
+        return TaskResult.builder(ExecutionStatus.SUCCEEDED).context(stageOutput).build();
+      }
+
+      // Retrieve artifacts
+      if (stageExecution.getContext().containsKey("expectedArtifacts")
+          && !((List) stageExecution.getContext().get("expectedArtifacts")).isEmpty()) {
+        try {
+          stageOutput.put("artifacts", JsonPath.parse(response.getBody()).read("artifacts"));
+        } catch (Exception e) {
+          webHookOutput.put(
+              "error", "Expected artifacts in webhook response couldn't be parsed " + e.toString());
+          return TaskResult.builder(ExecutionStatus.TERMINAL).context(stageOutput).build();
+        }
+      }
+      return TaskResult.builder(ExecutionStatus.SUCCEEDED).context(stageOutput).build();
+    } else {
+      webHookOutput.put("error", "The webhook request failed");
+      return TaskResult.builder(ExecutionStatus.TERMINAL).context(stageOutput).build();
+    }
+  }
+
+  private Object processResponseBodyAsJson(String responseBody) {
+    Object body = null;
+    try {
+      if (responseBody.startsWith("{")) {
+        body = objectMapper.readValue(responseBody, Map.class);
+      } else if (responseBody.startsWith("[")) {
+        body = objectMapper.readValue(responseBody, List.class);
+      }
+    } catch (JsonProcessingException ex) {
+      // Just leave body as string, probs not JSON
+      log.warn("Failed to parse webhook payload as JSON", ex);
+    }
```

Comment:
Don't have to fix it here, but this behavior is incorrect: It assumes that a valid JSON response must only be either a list or object, however according to the JSON spec, the root element can be of any of JSON's 7 types: https://www.json.org/json-en.html